File: 1008.Construct Binary Search Tree from Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def bstFromPreorder(self, preorder):
        """
        :type preorder: List[int]
        :rtype: TreeNode
        """
        dummy = TreeNode(float('inf'))
        path = [dummy]
        pointer = dummy
        for val in preorder:
            if val < pointer.val:
                pointer.left = TreeNode(val)
                path.append(pointer)
                pointer = pointer.left
            else:
                while val >= path[-1].val:
                    pointer = path.pop()
                if val < pointer.val:
                    pointer.left = TreeNode(val)
                    path.append(pointer)
                    pointer = pointer.left
                else:
                    pointer.right = TreeNode(val)
                    pointer = pointer.right
        return dummy.left
                

File: 1008.Construct Binary Search Tree from Preorder Traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {Integer[]} preorder
# @return {TreeNode}
def bst_from_preorder(preorder)
    @index = 0
    @array = preorder
    def dfs(smallest, biggest)
        if @index < @array.size && smallest <= @array[@index] && @array[@index] <= biggest
            root = TreeNode.new @array[@index]
            @index += 1
            root.left = dfs(smallest, [biggest, root.val].min)
            root.right = dfs([smallest, root.val].max, biggest)
            return root
        end
    end
    dfs(-1.0/0.0, 1.0/0.0)
end
